subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslCompilerPlugins.kt

```diff
@@ -39,14 +42,18 @@ class KotlinDslCompilerPlugins : Plugin<Project> {
         extensions.configure(SamWithReceiverExtension::class.java) { samWithReceiver ->
             samWithReceiver.annotation(HasImplicitReceiver::class.qualifiedName!!)
         }
+        plugins.apply(AssignmentSubplugin::class.java)
+        extensions.configure(AssignmentExtension::class.java) { assignment ->
+            assignment.annotation(SupportsKotlinAssignment::class.qualifiedName!!)
+        }
 
         afterEvaluate {
             kotlinDslPluginOptions {
                 tasks.withType<KotlinCompile>().configureEach {
                     it.kotlinOptions {
                         jvmTarget = this@kotlinDslPluginOptions.jvmTarget.get()
-                        apiVersion = "1.4"
-                        languageVersion = "1.4"
+                        apiVersion = "1.7"
+                        languageVersion = "1.7"
```

Comment:
Do we need to bump the language version for assignment overloading to work?

Reply:
No we don't. I just applied changes from https://github.com/gradle/gradle/pull/21910 that updates Kotlin to 1.7.

Reply:
Are we sure to switch to 1.7 in Gradle 8? In either case it seems prudent to separate those changes from this PR if possible.

Reply:
Yes, we will surely switch to 1.7 in Gradle 8. I will drop that change once 1.7 is merged to master. But I remembered I anyway need to upgrade API to at least 1.5 if this change is not applied, since with the 1.8 compiler you can build to minimum 1.5 language level.